src/Chapter2.hs

```diff
@@ -841,8 +873,11 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
-
+rotate :: Int -> [a] -> [a]
+rotate n l
+    | n < 0 = []
+    | otherwise = let len = length l
+                  in take len (drop n (cycle l))
```

Comment:
Unfortunately, `cycle` fails in runtime on empty lists :recycle: So you need to handle this case separately for this function to work properly.

Reply:
Actually I tested the empty list case and it works.

If I'm correct in the empty list case the last line evaluates to `take 0 (drop n (cycle []))` where take 0 directly returns `[]` and the `cycle` call gets never evaluated because of lazy evaluation.

But I agree explicitly catching that case is much nicer.